File: Regression/DecisionTreeRegression/decision_tree_regression.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Jun  8 09:27:02 2018

@author: rajchoudhary
"""

# Decision Tree Regression
# Non-linear and Non-continuous Regression model

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:-1].values
y = dataset.iloc[:, -1].values

# Fitting the Decision Tree Regression model to the dataset
from sklearn.tree import DecisionTreeRegressor
regressor = DecisionTreeRegressor(random_state = 0)
regressor.fit(X, y)

# Predicting the results using Decision Tree Regressor
y_pred = regressor.predict(6.5)

# Visualising the Decision Tree Regression model
# If we simply plot our regressor for the X dataset, then the plot will be continuous
# and the visualisation will be wrong.
# So we plot our visualisation for higher resolution
X_grid = np.arange(min(X), max(X), 0.01)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = 'red')
plt.plot(X_grid, regressor.predict(X_grid), color = 'blue')
plt.title('Bluff or Truth (Decision Tree Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

File: Regression/DecisionTreeRegression/decision_tree_regression.R
# Decision Tree Regression
# Non-linear and Non-continuous Regression Model

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Fitting the Decision Tree Regression model to the dataset
# install.packages('rpart')
library(rpart)
regressor = rpart(formula = Salary ~.,
                  data = dataset,
                  control = rpart.control(minsplit = 1))

# Predicting the results with the Decision Tree Regression model
y_pred = predict(regressor, newdata = data.frame(Level = 6.5))

# Visualising the Decision Tree Regression model with higher resolution
# install.packages('ggplot2')
library(ggplot2)
X_grid = seq(min(dataset$Level), max(dataset$Level), 0.01)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = X_grid, y = predict(regressor, newdata = data.frame(Level = X_grid))),
             colour = 'blue') +
  ggtitle('Truth or Bluff (Decision Tree Regression)') +
  xlab('Position Level') +
  ylab('Salary')
